Polish split bills

ErnWong opened the pull request:
Closes #129

**Description**

In the interest of time, the rest of the work needed for #129 are all bundled together into this PR.

Part 1/6 was already done: PR #149 

Part 2/6 - <kbd>[SHOW DIFF](https://github.com/se701-group6/A2/pull/155/files/e08b7f45886d55818b0214b76cf4e3956d00a349?diff=split&w=1)</kbd> - <kbd>[TRY DEMO ON CODESANDBOX.IO](https://bhj8f.csb.app/#/home/split)</kbd>
- Add a cross icon to each row, and hook it up to delete the row
- Add a drag handle to each row, and hook it up to [react-sortable-hoc](https://github.com/clauderic/react-sortable-hoc)
- There will always be at least two rows in the list. Deleting one of the two remaining rows will spawn a new row.

Part 3/6 - <kbd>[SHOW DIFF](https://github.com/se701-group6/A2/pull/155/files/e08b7f45886d55818b0214b76cf4e3956d00a349..d37b291af34922ec817ef565d75d320fedec1501?diff=split&w=1)</kbd> - <kbd>[TRY DEMO ON CODESANDBOX.IO](https://66li2.csb.app/#/home/split)</kbd>
- Style the page according to the new [HiFi prototypes](https://github.com/se701-group6/A2/wiki/HiFi-Prototypes#split-bill-form-updated). Loads Open Sans from Google Fonts.

Part 4/6 - <kbd>[SHOW DIFF](https://github.com/se701-group6/A2/pull/155/commits/0ef1ed7c41fcb90a1a21b61f5794c24cf05b643f?diff=split&w=1)</kbd> - <kbd>[TRY DEMO ON CODESANDBOX.IO](https://l0jgi.csb.app/#/home/split)</kbd>
- Animate list changes using [react-flip-toolkit](https://github.com/aholachek/react-flip-toolkit)

Part 5/6 - <kbd>[SHOW DIFF](https://github.com/se701-group6/A2/pull/155/files/0ef1ed7c41fcb90a1a21b61f5794c24cf05b643f..cca86f091a5979a14290e6e984883813cfd39a9b?diff=split&w=1)</kbd> - <kbd>[TRY DEMO ON CODESANDBOX.IO](https://2dh60.csb.app/#/home/split)</kbd>
- Replace `alert()`s with inline error validation (using `src/components/ErrorBubble/`)

Part 6/6 - <kbd>[SHOW DIFF](https://github.com/se701-group6/A2/pull/155/commits/6d6ebf8e371f07478c242b5d54227f09d314ae14)</kbd> - <kbd>[TRY DEMO ON CODESANDBOX.IO](https://tfvwo.csb.app/#/home/split)</kbd>
- Refactor folder structure (See #136)

Part 7/6 - <kbd>[SHOW DIFF](https://github.com/se701-group6/A2/pull/155/files/6d6ebf8e371f07478c242b5d54227f09d314ae14..e1d64dbc3274c0c44e63c713b21d898118fb7c05?diff=split&w=1)</kbd> - <kbd>[TRY DEMO ON CODESANDBOX.IO](https://6ziqd.csb.app/#/home/split)</kbd>
- Improved handling of edge cases like long inputs, whitespaces, rapid clicking

**Testing**

- Deleting first row, last row, middle row
- Deleting the first/second row when there are only two rows left
- Drag row to top / bottom / middle
- Deleting payer row
- Deleting payee row and checking that the error bubble appears
- Checking error bubbles appears when:
   - Total cost is zero / empty / negative
   - A name is empty / duplicate
- Deleting duplicate / empty row and checking that the error bubble disappears

**Checklist**
<!-- this section may be replaced with GitHub action checks later -->
- [x] I have tested my change
- [x] Code builds successfully and all tests pass
- [x] Docs have been added/updated if needed

ErnWong committed:
Add remove user handler in split bill page

ErnWong committed:
Add delete row button to split bill page

ErnWong committed:
Implement drag-n-drop reordering of split bill people

ErnWong committed:
Fix drag helper list style and add cursor for drag handle

ErnWong committed:
Style Split-Bill page according to new design mockups

- Open Sans added

ErnWong committed:
Allow placeholder bill title to be submitted

ErnWong committed:
Animate split-bill page when list changes

ErnWong committed:
Remove toggling behaviour for split bill toggle buttons

ErnWong committed:
Mark person's name required, and bill title not required

ErnWong committed:
Always show at least 2 rows in split bill form

ErnWong committed:
Implement split-bill inline error validation

- Introduction of ErrorBubble components
- Introduction of the errors object
- Create a "fake" progress feedback delay

ErnWong committed:
Don't reset payee when selected payer button is re-selected

The "Payer" button should only do something if that person is a "payee".
If the person is already a payer, pressing the "payer" button should not
clear someone else's payee status.

ErnWong committed:
Fix error bubble font weight

Bubble should use normal italics, but only light italics were downloaded
and the browser was forced to use artificial boldening.

ErnWong committed:
Fix bug where deleting a payee row keeps the payee

ErnWong committed:
Deleting last 2 rows should feel like "clearing" the rows

Previously:
Given only 2 rows in the list
When the first row is deleted
Then the second row moves up to the first row,
and a new row is added to the end to the second row position.

Now:
Given only 2 rows in the list
When the first row is deleted
Then a new row is added back into the first row,
and the second row is not affected.

This simplifies the user's impression of the animation.

ErnWong committed:
Refactor split-bill file structure to best practices

I.e. dedicated folder for Split, and ErrorBubble

ErnWong committed:
Prevent error bubble in row being overlapped by row below

This bug happened in inconsistent cases.
See comment in this commit.

ErnWong committed:
Trim whitespace off split bill textfields

ErnWong committed:
Fix fluid input not resizing with white spaces in input

wooneh commented on split-webapp/split/src/pages/Split/PeopleList.module.css:
```diff
@@ -0,0 +1,193 @@
+  min-width: 44px !important;
+  border-radius: 0 !important;
+  transition: color 0.3s !important;
+  color: #d6e3fe !important;
```
There seems to be an awful lot of `!important` in the css files, wasn't there a way to change the order of css injection/priority with material ui? Or is there some other way to avoid using `!important`?

ErnWong replied:
Yeah, I don't like the `!important` either. I'll see if there's a way to change the injection order.

ErnWong replied:
Yikes, using `<StyleProvider injectFirst>` to make material-ui's style lesser priority than our css appears to break a lot of the app's styles.

![image](https://user-images.githubusercontent.com/2609018/80162824-7e0fcd00-8628-11ea-989c-c604108f1a3a.png)

I think we should leave things as is for now. I'll create a new issue/story for this instead.

wooneh replied:
Sounds good!

wooneh commented on split-webapp/split/src/pages/Split/index.jsx:
```diff
@@ -0,0 +1,388 @@
+        errors.users[userId] = "Please enter a name";
+        errors.hasError = true;
+      } else if (uniqueNameSet.has(name)) {
+        errors.users[userId] = "Same name can’t appear twice";
```
Is there a reason you used 'RIGHT SINGLE QUOTATION MARK' (U+2019) here as opposed to 'APOSTROPHE' (U+0027)? Interested in how you ended up with it since the default is usually U+0027. I don't really have a preference either way unless someone else would like to comment.

ErnWong replied:
Yeah, I always preferred using the curly variants which looks more typographically correct, but now I wonder if it has the right semantics?

I double checked - it looks like Unicode version 13.0.0 recommends using U+2019 for contractions:

![image](https://user-images.githubusercontent.com/2609018/80158600-012b2600-861d-11ea-892c-69ce9c7b3472.png)

but [some people disagree with Unicode](https://tedclancy.wordpress.com/2015/06/03/which-unicode-character-should-represent-the-english-apostrophe-and-why-the-unicode-committee-is-very-wrong/) 🤔

ErnWong committed:
Gracefully handle large inputs in split bill form

A very long title should not push the total text field outside of the
box.

ErnWong committed:
Fix Safari focus stuck on previous textfield bug

ErnWong committed:
Flatten FluidInput's fake span structure

This is in attempt to simplify FluidInput's layout behaviour, hopefully
making it easier to debug CSS problems due to misalignments and
overflows.

ErnWong committed:
Allow header error to overflow outside, and fix FluidInput alignment

The cause of alignment problems was the padding we used to fix the
clipped descenders in the text fields.

The clipping of the descenders are now fixed by resetting the height and
line-height attributes.

ErnWong committed:
Reduce flip animation distortion due to border thickness

Distortion was noticeable on Safari, and in some cases, other browsers
too. To reproduce the distortion, create a list of people, then try
deleting a row. As you delete a row, you may notice that the header text
moves slightly even when it should remain still.

FrankWangma reviewed (approved):
I think everything looks fine to me. I would recommend having more than 2 people review this, since it is introducing very large changes to the code
